src/graph_ir.rs

```diff
@@ -71,34 +75,50 @@ pub fn compile(
     nodes: impl IntoIterator<Item = Node>,
     edges: impl IntoIterator<Item = Edge>,
 ) -> CompiledSchedule {
-    GraphIR::preprocess(num_port_types, nodes, edges)
+    let nodes: FnvHashMap<NodeID, Node> = nodes.into_iter().map(|n| (n.id, n)).collect();
+    let mut adjacent: FnvHashMap<NodeID, NodeEdges> = FnvHashMap::default();
+    for edge in edges.into_iter() {
+        let src = adjacent.entry(edge.src_node).or_insert_with(NodeEdges::new);
+        src.outgoing.push(edge);
+        let dst = adjacent.entry(edge.dst_node).or_insert_with(NodeEdges::new);
+        dst.incoming.push(edge);
+    }
+
+    GraphIR::start(num_port_types, &nodes, &adjacent)
         .sort_topologically()
         .solve_latency_requirements()
         .solve_buffer_requirements()
         .merge()
 }
 
-impl GraphIR {
-    /// Construct a [GraphIR] instance from lists of nodes and edges, building
-    /// up the adjacency table and creating an empty schedule.
-    pub fn preprocess(
+/// Compile a graph from a pre-existing lists of nodes and their edges.
+///
+/// It is up to the user to uphold the correctness of these lists.
+pub fn compile_preprocessed(
+    num_port_types: usize,
+    nodes: &FnvHashMap<NodeID, Node>,
+    adjacent: &FnvHashMap<NodeID, NodeEdges>,
+) -> CompiledSchedule {
+    GraphIR::start(num_port_types, nodes, adjacent)
+        .sort_topologically()
+        .solve_latency_requirements()
+        .solve_buffer_requirements()
+        .merge()
+}
+
+impl<'a> GraphIR<'a> {
+    /// Construct a [GraphIR] instance from a pre-existing lists of nodes and
+    /// their edges.
+    ///
+    /// It is up to the user to uphold the correctness of these lists.
+    pub fn start(
         num_port_types: usize,
-        nodes: impl IntoIterator<Item = Node>,
-        edges: impl IntoIterator<Item = Edge>,
+        nodes: &'a FnvHashMap<NodeID, Node>,
+        adjacent: &'a FnvHashMap<NodeID, NodeEdges>,
```

Comment:
Like mentioned above, this is useful but the GraphIR should maintain the whole state.